text/3735-field-projections.md

```diff
@@ -0,0 +1,1344 @@
+    Instead we use an external tool called [`klint`] for that purpose.
+
+[`klint`]: https://rust-for-linux.com/klint
+
+```rust
+mod rcu {
+    pub struct Rcu<P> {
+        inner: UnsafeCell<P>,
+        // we require this to opt-out of uniqueness of `&mut`.
+        // if `UnsafePinned` were available, we would use that instead.
+        _phantom: PhantomPinned,
+    }
+    
+    impl<P: Deref> Rcu<P> {
+        pub fn read<'a>(&'a self, _guard: &'a RcuGuard) -> &'a P::Target;
+        pub fn set(self: Pin<&mut Self>, new: P) -> Old<P>;
+    }
+
+    pub struct Old<P>(/* ... */);
+    
+    impl<P> Drop for Old<P> {
+        fn drop() {
+            unsafe { bindings::synchronize_rcu() };
+        }
+    }
+}
+```
+
+The `Old` type is responsible for calling `synchronize_rcu` before dropping the old value.
+
+Note that `set` takes a pinned mutable reference to `Rcu`. This is important, since it might not be
+obvious why there is pinning involved here. Firstly, we need to take a mutable reference, since
+writers still need to be synchronized. Secondly, since there are still concurrent shared references,
+we must not allow users to use `mem::swap`, since that would change the value without the required
+compiler and CPU barriers in place.
+
+Now to the crux of the issue and why field projection comes up here: A common use-case of RCU is to
+protect data inside of a struct that is itself protected by a lock. Since the data is protected by
+RCU, we don't need to hold the lock to read the data. However, locks do not allow access to the
+inner value without locking it (that's kind of their whole point...). So we need a way to get to the
+`Rcu<P>` without locking the lock. Using field projection, we would allow projections for fields of
+type `Rcu` from `&Lock` to `&Rcu<P>`.
+
+This way, readers can use field projection and the `Rcu::read` function and writers can continue to
+lock the lock and then use `Rcu::set`.
+
+#### RCU API Usage Examples
+
+```rust
+struct BufferConfig {
+    flush_sensitivity: u8,
+}
+
+struct Buffer {
+    // We also require `Rcu` to be pinned, because `&mut Rcu` must not exist (otherwise one could
+    // call mem::swap).
+    #[pin]
+    cfg: Rcu<Box<BufferConfig>>,
+    buf: Vec<u8>,
+}
+
+struct MyDriver {
+    // The `Mutex` in the kernel needs to be pinned.
+    #[pin]
+    buf: Mutex<Buffer>,
+}
+```
+
+Here the struct that is protected by the lock is `Buffer` and the data that is protected by RCU
+inside of this struct is `BufferConfig`. To read the config, we now don't have to lock the lock,
+instead we can read it using field projection:
+
+```rust
+impl MyDriver {
+    fn buffer_config<'a>(&'a self, rcu_guard: &'a RcuGuard) -> &'a BufferConfig {
+        let buf: &Mutex<Buffer> = &self.buf;
+        // Here we use the special projections set up for `Mutex` with fields of type `Rcu<T>`.
+        let cfg: &Rcu<Box<BufferConfig>> = buf->cfg;
```

Comment:
I will add it, it is our own mutex type, so we shouldn't hit orphan rules. Also, in RfL we want to relax orphan rules anyways, so for us it might not be an issue.

Reply:
Originally, I intended to add the code, but I forgot. I now noticed an issue with my idea for how this would work, since in the meantime, I added `impl<'a, T, F: Field<Base = T>> Project<F> for &'a T`. That overlaps with the following:
```rust



impl<'a, T> Projectable for &'a Mutex<T> {
    type Inner = T;
}

impl<'a, T, F, U: 'a> Project<F> for &'a Mutex<T>
where
    F: Field<Base = T, Type = Rcu<U>>
{
    type Output = &'a Rcu<U>;

    fn project(self) -> Self::Output {
        todo!()
    }
}
```

- removing the blanket impl for all `&'a T` can't be done, because we need it to be able to project `&UnsafeCell` and other types with `#[projecting]`.
- this looks like something specialization might help with, but I am totally out of the loop wrt that, last time there wasn't much movement there, so I think we have to come up with something else.
- one thing that we could do is the following:
give containers with `#[projecting]` the ability to select which fields are passed through, then for `Mutex`, we could only allow `Rcu<U>` fields.
I haven't thought about this for long, maybe this is unsound in our case, so I'm not too inclined to jump on this, but maybe with more thought, it ends up a good idea...

Reply:
I see, yeah having a separate reference type for a mutex that can be projected might be possible. But it also is a bit annoying...

(Note that the `MutexProjection` type that you defined should implement `Projectable`/`Project` instead of being `#[projecting]`)

Reply:
I suspect this is a general issue with `project` taking `self`: likely other primitives will want to be able to project from a shared/mutable ref to the thing, not just the thing, e.g. Pin might want to take a `&mut Pin<&mut T>` so you can reuse the pointer afterwards. You may need to separate `ProjectRef`/`ProjectMut`/`ProjectMove` like `Deref` had to for likely similar reasons.

Reply:
We could do like for `Deref`: it wouldn't be `&T: ProjectMove`, it would be `&T: ProjectRef`, which doesn't conflict with `Mutex<T>: ProjectRef`.

Reply:
I'm confused, how exactly would that look like? Is `ProjectMove` the current version of `Project` and
```rust
pub trait ProjectRef<F: Field<Base = Self::Inner>>: Projectable {
    type Output<'a>;
    fn project(&self) -> Self::Output<'_>;
}
```
and `ProjectMut` is the same, but with `&mut self` as the receiver?

What if `&Foo` implements `ProjectMove` and `Foo` implements `ProjectRef`? What does `->` do in that case?

Reply:
Hm, I didn't realize this would require a GAT.  I haven't thought this through but I'd imagine when there's conflict we'd have to figure out a rule to infer the right one to use, like `Deref` vs `DerefMut` do today.

Reply:
If you have a `ProjectibleBase` trait, then you could do  `impl<T> !ProjectibleBase for Mutex<T>`.

Tho it would be annoying for the implementor of Mutex since they'll have to manually do field accesses with
```
fn inner_mutex(&self) -> &InnerMutex {
    field_of!(Self, inner).project_ref(self)
}
```